### DISCUSSION TITLE
Problems with typing of `error.format()` of discriminated union

### DISCUSSION THREAD

**Discussion by hanayashiki** [colinhacks/zod#2765] [Q&A] [answered]
<img width="700" alt="image" src="https://github.com/colinhacks/zod/assets/26056783/96ffdf84-161c-4a61-827f-10b0cc22082c">

Here we define a discriminated union, and try to get the typing of `error.format()`. However, the typing seems broken and we cannot get to the error of each case. Also, `f.tag` is typed as `{}`, which is too general.

<img width="531" alt="image" src="https://github.com/colinhacks/zod/assets/26056783/c8cc1eba-6249-42e2-9aa3-82034f022269">

Reproduction on codesandbox:

https://codesandbox.io/s/problems-with-typing-of-error-format-of-discriminated-union-l2c68d?file=/index.ts

**Comment by JacobWeisenburger** [comment, marked as answer]
This is a limitation of TypeScript. If you have a union of 2 object types, then you can't know which properties exist until you check the discriminator. But zod doesn't check the discriminator until runtime, so it can't know which properties exist at compile time in the editor.

![image](https://github.com/colinhacks/zod/assets/31667350/b4546013-bfae-42bf-a0f8-3cf822e6717e)

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
